BCVaccineValidator/Classes/Services/RevocationManager.swift

```diff
@@ -0,0 +1,291 @@
+//
+//  RevocationManager.swift
+//  BCVaccineValidator
+//
+//  Created by Mohamed Afsar on 15/02/22.
+//
+
+import Foundation
+import CoreMedia
+
+internal final class RevocationManager: DirectoryManager {
+    static let shared = RevocationManager()
+    
+    private typealias Issuer = String
+    private typealias IssuerFilePathSafeName = String
+    private typealias KeyId = String
+    private typealias RevocationDataSet = SynchronizedDictionary<IssuerFilePathSafeName, SynchronizedDictionary<KeyId, RevocationData>>
+    
+    private static let crlFileNameFormat = "%@.json" // NO I18N
+    private static let crlDirectoryName = "crl" // NO I18N
+    private static let vaxValidatorCrlDataUDKeyFormat = "vaxValidatorCrlData_%@_%@_savedAt" // NO I18N
+    
+    private lazy var crlDirectoryURL = documentDirectory().appendingPathComponent(RevocationManager.crlDirectoryName)
+    private let fetchingKeyIds = SynchronizedSet<String>()
+    private let revocationDataSet = RevocationDataSet()
+    private let notFoundEndPoints = SynchronizedDictionary<IssuerFilePathSafeName, SynchronizedDictionary<KeyId, Date>>()
+    
+    init() {
+        seedLocalRevocationData(to: revocationDataSet)
+    }
+}
+
+// MARK: Functions
+internal extension RevocationManager {
+    func isPvcRevoked(issuer: String, issueDate: Date, rId: String, kId: String) -> Bool {
+        downloadAndCacheIfNeeded(completion: { })
+        // Reject the Health Card if the calculated rid is contained in the CRL's rids array and (if a timestamp suffix is present) the Health Card’s nbf is value is before the timestamp.
+        guard let revocationData = revocationDataSet[getFilePathSafeName(issuer: issuer)]?[kId],
+              revocationData.rids.contains(rId) else {
+            return false
+        }
+        if let associatedDate = revocationData.revocationExpiry[rId] {
+            return issueDate < associatedDate
+        } else {
+            return true
+        }
+    }
+    
+    func downloadAndCacheIfNeeded(completion: @escaping () -> Void) {
+        // Get list of issuers
+        IssuerManager.shared.getIssuers { [weak self] result in
+            guard let self = self else { return }
+            guard let issuers = result else {
+                // It is highly unlikely that result is nil.
+                Logger.logFailure("Critical Error: No issuers found") // NO I18N
+                return completion()
+            }
+            let displatchGroup = DispatchGroup()
```

Comment:
there is a typo here, please rename "displatchGroup" to "dispatchGroup"

Reply:
Will fix